### PR TITLE
add entity claim references

### PR DESCRIPTION
**Pull request by maxlath**
Addressing #710

This implementation lets 2 claim shapes co-exist:
* The classic standalone claim value (aka "simplified claim"):
```js
{
  'wdt:P2969': [ '51324755' ]
}
```
* The new claim object (aka "expanded claim"):
```js
{
  'wdt:P2969': [
    {
      value: '51324755',
      references: [
        {
          'wdt:P648': ['OL6089177M'],
          'wdt:P813': ['2024-05-11']
        }
      ]
    }
  ]
}
```

This allows to transition from the existing entities to the new expanded format without having to rewrite all entities and their corresponding patches, but if this claim format coexistence is deemed too costly, such a transition could be done.

This PR can be broken down into several subparts:
* update all pieces of logic that needed to access entities claims values
* get entities with/without references
* create entities with references (currently only needed by tests)
* set references in authorities seeds

That's a fat PR as this data model change means every part of the codebase trying to access or manipulate claim values now needs to go through an adaptor to parse the possible claim objects. The changes are limited by keeping the non-expanded claim format as default format in serialized entities:
* Only code manipulating non-serialized entity docs has to use adaptors
* As the `getEntitiesByUris` function returns serialized entities, all the code outside the entities' realm should preferably use it, to be saved from claim object related considerations.
* This PR doesn't require any change client-side.

Existing tests pass, but some new tests might be missing.

### CODE REVIEW

File: server/controllers/entities/lib/validate_and_format_claim.ts [hunk 1]
```diff
@@ -1,8 +1,49 @@
...
+    validateAndFormatReferences(newClaim)
+  }
   return validateAndFormatClaimValue(params)
 }
+
+function validateAndFormatReferences (claim: InvClaimObject) {
+  const { references } = claim
+  if (!references) return
+  if (!(references instanceof Array)) {
+    throw newError('invalid reference array', 400, { claim, references })
```
Comment (jum-s) [inline server/controllers/entities/lib/validate_and_format_claim.ts:24]: Error message could be more explicit imo:

```suggestion
    throw newError('invalid references value, should be an array', 400, { claim, references })
```

File: server/controllers/items/lib/export/add_entities_data.ts [hunk 2]
```diff
@@ -12,15 +12,15 @@ export default async function (item) {
...
+    item.publisherUri = entity.claims['wdt:P123']?.[0]
     if (item.publisherUri) {
       item.publisher = await getEntityByUri({ uri: item.publisherUri })
     }
     item.translatorsUris = entity.claims['wdt:P655']
     if (item.translatorsUris) {
       item.translators = await getEntitiesList(item.translatorsUris)
     }
-    item.editionLangUri = entity.claims['wdt:P407'] && entity.claims['wdt:P407'][0]
+    item.editionLangUri = entity.claims['wdt:P407']?.[0]
```
[the remark this answers is not in the document]
Comment (maxlath) [reply server/controllers/items/lib/export/add_entities_data.ts:23]: `getFirstClaimValue` is required for getting a claim value when there might be both direct value claims and claim objects, which is typically the case when manipulating the entity doc pre-serialization. `getEntityByUri` functions return serialized entities, so either only claim objects (if called with `{ expanded: true }`) or only direct claim values (as is the case in the code above). In both cases, `getFirstClaimValue` should still work but isn't needed, as we already know the claims shape.

File: server/models/entity.ts [hunk 3]
```diff
@@ -295,3 +321,99 @@ function deleteLabel (doc, lang) {
...
+  }
+}
+
+function includesReference (references: Reference[], reference: Reference) {
+  return references.find(ref => sameObjects(ref, reference))
+}
+
+// The option is named 'keepReferences' rather than 'includeReferences' to make
+// the function options compatible with wikibase-sdk simplifyClaims options
+export function simplifyInvClaims (claims: Claims, { keepReferences = false } = {}) {
```
[the remark this answers is not in the document]
Comment (maxlath) [reply server/models/entity.ts]: done in 31b52f464

File: server/controllers/entities/lib/validate_and_format_claim.ts [hunk 4]
```diff
@@ -1,8 +1,49 @@
...
+}
+
+function validateAndFormatReference (reference: unknown, claim: InvClaimObject) {
+  if (!isNonEmptyPlainObject(reference)) {
+    throw newError('invalid reference', 400, { reference, claim })
+  }
+  for (const [ property, values ] of objectEntries(reference)) {
+    validateProperty(property)
+    if (!(arrayIncludes(allowlistedReferenceProperties, property))) {
+      throw newError("This property isn't allowed in reference snaks", 400, { property, allowlistedReferenceProperties })
```
[the remark this answers is not in the document]
Comment (maxlath) [reply server/controllers/entities/lib/validate_and_format_claim.ts]: reworded in 4cd0f6750 (I prefered the form " in a reference" as a reference may contain several properties)

File: server/controllers/entities/lib/validate_and_format_claim.ts (hunk 1 above)
Comment (maxlath) [reply server/controllers/entities/lib/validate_and_format_claim.ts]: done in b509db6f8

File: server/controllers/items/lib/export/add_entities_data.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jum-s) [reply server/controllers/items/lib/export/add_entities_data.ts:23]: Yes, this comment was only made for the purpose of having the same function to access some data.
[non-blocking]

**Review by jum-s — APPROVED** [review]

**Pull request merged**